File: language/typescript/fib.ts
function fib(n: number): number {
  return n < 2 ? n : fib(n - 1) + fib(n - 2);
}

function fibLinear(n: number, prevFib: number = 0, fib: number = 1): number {
  if (n < 2) return n;
  while (--n) [prevFib, fib] = [fib, prevFib + fib];
  return fib;
}

function fibFormula(n: number): number {
  return Math.round(((5 ** 0.5 + 1) / 2) ** n / 5 ** 0.5);
}

function fibTailRecursive(n: number, prevFib: number = 0, fib: number = 1): number {
  return n === 0 ? prevFib : fibTailRecursive(n - 1, fib, prevFib + fib);
}

const input: number = typeof process !== "undefined" && (parseInt(process.argv[2], 10) || 29);
console.log(`Fibonacci ${input}: ${fibLinear(input)}`);

File: language/scala/fib.scala
object Fibonacci {
  def fib(n: Int): Int =
    if (n < 2) n else fib(n - 1) + fib(n - 2);

  def fibLinear(n: Int): Int = {
    if (n < 2) return n;
    var prevFib = 0
    var fib = 1
    (2 to n) foreach (x => {
      val temp = prevFib + fib;
      prevFib = fib;
      fib = temp;
    })
    return fib;
  }

  def fibFormula(n: Int): Int =
    math.round(math.pow((math.pow(5, 0.5) + 1) / 2, n).toFloat / math.pow(5, 0.5).toFloat)

  def fibTailRecursive(n: Int, prevFib: Int = 0, fib: Int = 1): Int =
    if (n == 0) prevFib else fibTailRecursive(n - 1, fib, prevFib + fib);

  def fibFormal(n: Int): Int = n match {
    case 0 => 0
    case 1 => 1
    case _ => fibFormal(n - 1) + fibFormal(n - 2)
  }

  def main(args: Array[String]): Unit = {
    val input = if (args.length > 0) args(0).toInt else 29
    println("Fibonacci " + input + ": " + fibLinear(input))
  }
}
